File: frontend/src/app/services/project-ownership.service.ts
import { Injectable } from '@angular/core';
import { Project } from '../models/pusharoo.models';

@Injectable({ providedIn: 'root' })
export class ProjectOwnershipService {
  canManage(project: Project | null | undefined, walletAddress: string): boolean {
    const creatorAddress = project?.createdByWalletAddress?.trim();

    return Boolean(creatorAddress) && creatorAddress === walletAddress.trim();
  }

  managementError(project: Project | null | undefined, walletAddress: string): string {
    if (!walletAddress.trim()) {
      return 'Connect the project creator wallet before continuing.';
    }

    if (!project?.createdByWalletAddress?.trim()) {
      return 'This legacy project has no verified creator. Recover ownership before managing it.';
    }

    return this.canManage(project, walletAddress)
      ? ''
      : 'Only the project creator can manage versions and deployments.';
  }
}

File: backend/Services/ProjectOwnershipService.cs
using backend.Models;

namespace backend.Services;

public sealed class ProjectOwnershipService
{
    public ProjectOwnershipValidationResult ValidateCreatorRecord(ProjectDocument project)
    {
        return string.IsNullOrWhiteSpace(project.CreatedByWalletAddress)
            || string.IsNullOrWhiteSpace(project.CreatedByWalletScriptHash)
            || string.IsNullOrWhiteSpace(project.CreatedByWalletPublicKey)
            ? Fail("Project ownership cannot be verified. This legacy project is read-only until ownership is recovered.")
            : ProjectOwnershipValidationResult.Valid;
    }

    public ProjectOwnershipValidationResult ValidateCanManage(
        ProjectDocument project,
        string? walletAddress)
    {
        var creatorRecord = ValidateCreatorRecord(project);
        if (!creatorRecord.IsValid)
        {
            return creatorRecord;
        }

        if (string.IsNullOrWhiteSpace(walletAddress))
        {
            return Fail("Wallet address is required.");
        }

        var creatorWalletAddress = project.CreatedByWalletAddress;
        if (string.IsNullOrWhiteSpace(creatorWalletAddress))
        {
            return Fail("Project ownership cannot be verified. This legacy project is read-only until ownership is recovered.");
        }

        return string.Equals(
            creatorWalletAddress.Trim(),
            walletAddress.Trim(),
            StringComparison.Ordinal)
            ? ProjectOwnershipValidationResult.Valid
            : Fail("Only the project creator can manage versions and deployments.");
    }

    private static ProjectOwnershipValidationResult Fail(string error)
    {
        return new ProjectOwnershipValidationResult(false, error);
    }
}

public sealed record ProjectOwnershipValidationResult(bool IsValid, string Error)
{
    public static ProjectOwnershipValidationResult Valid { get; } = new(true, string.Empty);
}
